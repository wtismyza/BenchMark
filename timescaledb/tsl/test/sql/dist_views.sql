-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

---------------------------------------------------
-- Test views and size_utils functions on distributed hypertable
---------------------------------------------------
\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER

SET client_min_messages TO ERROR;
DROP DATABASE IF EXISTS view_node_1;
DROP DATABASE IF EXISTS view_node_2;
DROP DATABASE IF EXISTS view_node_3;
SELECT * FROM add_data_node('view_node_1', host => 'localhost',
                            database => 'view_node_1');
SELECT * FROM add_data_node('view_node_2', host => 'localhost',
                            database => 'view_node_2');
SELECT * FROM add_data_node('view_node_3', host => 'localhost',
                            database => 'view_node_3');

GRANT USAGE ON FOREIGN SERVER view_node_1, view_node_2, view_node_3 TO :ROLE_1;
SET client_min_messages TO NOTICE;
SET ROLE :ROLE_1;
SELECT setseed(1);

CREATE TABLE dist_table(time timestamptz NOT NULL, device int, temp float, timedim date NOT NULL);
SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);
SELECT add_dimension('dist_table', 'timedim', chunk_time_interval=>'7 days'::interval);
INSERT INTO dist_table SELECT t, (abs(timestamp_hash(t::timestamp)) % 10) + 1, 80, '2020-01-01'
FROM generate_series('2018-03-02 1:00'::TIMESTAMPTZ, '2018-03-04 1:00', '1 hour') t;
ALTER TABLE dist_table SET (timescaledb.compress, timescaledb.compress_segmentby='device', timescaledb.compress_orderby = 'time DESC');

-- Test views with compression 
BEGIN;
SELECT compress_chunk(chunk)
FROM show_chunks('dist_table') AS chunk
ORDER BY chunk
LIMIT 1;

SELECT * FROM timescaledb_information.hypertables
WHERE hypertable_name = 'dist_table';
SELECT * from timescaledb_information.chunks 
ORDER BY hypertable_name, chunk_name;
SELECT * from timescaledb_information.dimensions 
ORDER BY hypertable_name, dimension_number;

SELECT * FROM chunks_detailed_size('dist_table'::regclass) 
ORDER BY chunk_name, node_name;
SELECT * FROM hypertable_detailed_size('dist_table'::regclass)
ORDER BY node_name;;

---tables with special characters in the name ----
CREATE TABLE "quote'tab" ( a timestamp,  b integer);
SELECT create_distributed_hypertable( '"quote''tab"', 'a', 'b', replication_factor=>2, chunk_time_interval=>INTERVAL '1 day');
INSERT into "quote'tab" select generate_series( '2020-02-02 10:00', '2020-02-05 10:00' , '1 day'::interval), 10;
SELECT * FROM  chunks_detailed_size( '"quote''tab"') ORDER BY chunk_name, node_name;

CREATE TABLE "special#tab" ( a timestamp,  b integer);
SELECT create_hypertable( 'special#tab', 'a', 'b', replication_factor=>2, chunk_time_interval=>INTERVAL '1 day');
INSERT into "special#tab" select generate_series( '2020-02-02 10:00', '2020-02-05 10:00' , '1 day'::interval), 10;
SELECT * FROM  chunks_detailed_size( '"special#tab"') ORDER BY chunk_name, node_name;
SELECT * FROM  hypertable_index_size( 'dist_table_time_idx') ;
