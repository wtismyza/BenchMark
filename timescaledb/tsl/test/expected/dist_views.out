-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
---------------------------------------------------
-- Test views and size_utils functions on distributed hypertable
---------------------------------------------------
\c :TEST_DBNAME :ROLE_CLUSTER_SUPERUSER
SET client_min_messages TO ERROR;
DROP DATABASE IF EXISTS view_node_1;
DROP DATABASE IF EXISTS view_node_2;
DROP DATABASE IF EXISTS view_node_3;
SELECT * FROM add_data_node('view_node_1', host => 'localhost',
                            database => 'view_node_1');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 view_node_1 | localhost | 55432 | view_node_1 | t            | t                | t
(1 row)

SELECT * FROM add_data_node('view_node_2', host => 'localhost',
                            database => 'view_node_2');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 view_node_2 | localhost | 55432 | view_node_2 | t            | t                | t
(1 row)

SELECT * FROM add_data_node('view_node_3', host => 'localhost',
                            database => 'view_node_3');
  node_name  |   host    | port  |  database   | node_created | database_created | extension_created 
-------------+-----------+-------+-------------+--------------+------------------+-------------------
 view_node_3 | localhost | 55432 | view_node_3 | t            | t                | t
(1 row)

GRANT USAGE ON FOREIGN SERVER view_node_1, view_node_2, view_node_3 TO :ROLE_1;
SET client_min_messages TO NOTICE;
SET ROLE :ROLE_1;
SELECT setseed(1);
 setseed 
---------
 
(1 row)

CREATE TABLE dist_table(time timestamptz NOT NULL, device int, temp float, timedim date NOT NULL);
SELECT create_distributed_hypertable('dist_table', 'time', 'device', replication_factor => 2);
 create_distributed_hypertable 
-------------------------------
 (1,public,dist_table,t)
(1 row)

SELECT add_dimension('dist_table', 'timedim', chunk_time_interval=>'7 days'::interval);
          add_dimension          
---------------------------------
 (3,public,dist_table,timedim,t)
(1 row)

INSERT INTO dist_table SELECT t, (abs(timestamp_hash(t::timestamp)) % 10) + 1, 80, '2020-01-01'
FROM generate_series('2018-03-02 1:00'::TIMESTAMPTZ, '2018-03-04 1:00', '1 hour') t;
ALTER TABLE dist_table SET (timescaledb.compress, timescaledb.compress_segmentby='device', timescaledb.compress_orderby = 'time DESC');
-- Test views with compression 
BEGIN;
SELECT compress_chunk(chunk)
FROM show_chunks('dist_table') AS chunk
ORDER BY chunk
LIMIT 1;
               compress_chunk                
---------------------------------------------
 _timescaledb_internal._dist_hyper_1_1_chunk
(1 row)

SELECT * FROM timescaledb_information.hypertables
WHERE hypertable_name = 'dist_table';
 hypertable_schema | hypertable_name |    owner    | num_dimensions | num_chunks | compression_enabled | is_distributed | replication_factor |              data_nodes               | tablespaces 
-------------------+-----------------+-------------+----------------+------------+---------------------+----------------+--------------------+---------------------------------------+-------------
 public            | dist_table      | test_role_1 |              3 |          3 | f                   | t              |                  2 | {view_node_1,view_node_2,view_node_3} | 
(1 row)

SELECT * from timescaledb_information.chunks 
ORDER BY hypertable_name, chunk_name;
 hypertable_schema | hypertable_name |     chunk_schema      |      chunk_name       | primary_dimension |  primary_dimension_type  |         range_start          |          range_end           | range_start_integer | range_end_integer | is_compressed | chunk_tablespace |        data_nodes         
-------------------+-----------------+-----------------------+-----------------------+-------------------+--------------------------+------------------------------+------------------------------+---------------------+-------------------+---------------+------------------+---------------------------
 public            | dist_table      | _timescaledb_internal | _dist_hyper_1_1_chunk | time              | timestamp with time zone | Wed Feb 28 16:00:00 2018 PST | Wed Mar 07 16:00:00 2018 PST |                     |                   | f             |                  | {view_node_1,view_node_2}
 public            | dist_table      | _timescaledb_internal | _dist_hyper_1_2_chunk | time              | timestamp with time zone | Wed Feb 28 16:00:00 2018 PST | Wed Mar 07 16:00:00 2018 PST |                     |                   | f             |                  | {view_node_2,view_node_3}
 public            | dist_table      | _timescaledb_internal | _dist_hyper_1_3_chunk | time              | timestamp with time zone | Wed Feb 28 16:00:00 2018 PST | Wed Mar 07 16:00:00 2018 PST |                     |                   | f             |                  | {view_node_1,view_node_3}
(3 rows)

SELECT * from timescaledb_information.dimensions 
ORDER BY hypertable_name, dimension_number;
 hypertable_schema | hypertable_name | dimension_number | column_name |       column_type        | dimension_type | time_interval | integer_interval | integer_now_func | num_partitions 
-------------------+-----------------+------------------+-------------+--------------------------+----------------+---------------+------------------+------------------+----------------
 public            | dist_table      |                1 | time        | timestamp with time zone | Time           | @ 7 days      |                  |                  |               
 public            | dist_table      |                2 | device      | integer                  | Space          |               |                  |                  |              3
 public            | dist_table      |                3 | timedim     | date                     | Time           | @ 7 days      |                  |                  |               
(3 rows)

SELECT * FROM chunks_detailed_size('dist_table'::regclass) 
ORDER BY chunk_name, node_name;
     chunk_schema      |      chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |  node_name  
-----------------------+-----------------------+-------------+-------------+-------------+-------------+-------------
 _timescaledb_internal | _dist_hyper_1_1_chunk |        8192 |       32768 |        8192 |       49152 | view_node_1
 _timescaledb_internal | _dist_hyper_1_1_chunk |        8192 |       32768 |        8192 |       49152 | view_node_2
 _timescaledb_internal | _dist_hyper_1_2_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_1_2_chunk |        8192 |       32768 |           0 |       40960 | view_node_3
 _timescaledb_internal | _dist_hyper_1_3_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_1_3_chunk |        8192 |       32768 |           0 |       40960 | view_node_3
(6 rows)

SELECT * FROM hypertable_detailed_size('dist_table'::regclass)
ORDER BY node_name;;
 table_bytes | index_bytes | toast_bytes | total_bytes |  node_name  
-------------+-------------+-------------+-------------+-------------
       16384 |       65536 |        8192 |       90112 | view_node_1
       16384 |       65536 |        8192 |       90112 | view_node_2
       16384 |       65536 |           0 |       81920 | view_node_3
(3 rows)

---tables with special characters in the name ----
CREATE TABLE "quote'tab" ( a timestamp,  b integer);
SELECT create_distributed_hypertable( '"quote''tab"', 'a', 'b', replication_factor=>2, chunk_time_interval=>INTERVAL '1 day');
NOTICE:  adding not-null constraint to column "a"
 create_distributed_hypertable 
-------------------------------
 (2,public,quote'tab,t)
(1 row)

INSERT into "quote'tab" select generate_series( '2020-02-02 10:00', '2020-02-05 10:00' , '1 day'::interval), 10;
SELECT * FROM  chunks_detailed_size( '"quote''tab"') ORDER BY chunk_name, node_name;
     chunk_schema      |      chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |  node_name  
-----------------------+-----------------------+-------------+-------------+-------------+-------------+-------------
 _timescaledb_internal | _dist_hyper_2_4_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_2_4_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_2_5_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_2_5_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_2_6_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_2_6_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_2_7_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_2_7_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
                       |                       |             |             |             |             | view_node_3
(9 rows)

CREATE TABLE "special#tab" ( a timestamp,  b integer);
SELECT create_hypertable( 'special#tab', 'a', 'b', replication_factor=>2, chunk_time_interval=>INTERVAL '1 day');
NOTICE:  adding not-null constraint to column "a"
    create_hypertable     
--------------------------
 (3,public,special#tab,t)
(1 row)

INSERT into "special#tab" select generate_series( '2020-02-02 10:00', '2020-02-05 10:00' , '1 day'::interval), 10;
SELECT * FROM  chunks_detailed_size( '"special#tab"') ORDER BY chunk_name, node_name;
     chunk_schema      |       chunk_name       | table_bytes | index_bytes | toast_bytes | total_bytes |  node_name  
-----------------------+------------------------+-------------+-------------+-------------+-------------+-------------
 _timescaledb_internal | _dist_hyper_3_10_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_3_10_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_3_11_chunk |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_3_11_chunk |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_3_8_chunk  |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_3_8_chunk  |        8192 |       32768 |           0 |       40960 | view_node_2
 _timescaledb_internal | _dist_hyper_3_9_chunk  |        8192 |       32768 |           0 |       40960 | view_node_1
 _timescaledb_internal | _dist_hyper_3_9_chunk  |        8192 |       32768 |           0 |       40960 | view_node_2
                       |                        |             |             |             |             | view_node_3
(9 rows)

SELECT * FROM  hypertable_index_size( 'dist_table_time_idx') ;
 hypertable_index_size 
-----------------------
                 81920
(1 row)

